functionary/builder/utils.py

```diff
@@ -33,34 +34,49 @@ def extract_package_definition(package_contents: bytes) -> dict:
 
     Returns:
         The package definition yaml loaded as a dict
+
+    Raises:
+        InvalidPackage: Raised whenever an error is encountered while extracting
+            package contents from the package.yaml in the tarball
     """
     package_contents_io = io.BytesIO(package_contents)
 
     try:
         tarball = tarfile.open(fileobj=package_contents_io, mode="r")
+        package_definition = _extract_package_definition(tarball)
+        return package_definition
     except tarfile.ReadError:
         raise InvalidPackage(
             "Could not untar package file. Make sure it is a valid gzipped tarball."
         )
-
-    def close_files():
-        package_contents_io.close()
-        tarball.close()
-
-    try:
-        package_definition_io = tarball.extractfile("package.yaml")
     except KeyError:
-        close_files()
         raise InvalidPackage("package.yaml not found")
+    except (InvalidPackage, YAMLError, ParserError):
+        raise InvalidPackage("package.yaml is invalid YAML")
+    finally:
+        tarball.close()
+        package_contents_io.close()
+
 
-    if package_definition_io is None:
-        close_files()
-        raise InvalidPackage("package.yaml found, but is not a regular file")
+def _extract_package_definition(tarball: tarfile.TarFile) -> dict:
+    """Extract the package definition from given tarfile
+
+    Args:
+        tarball: The opened tarfile in read mode
 
-    package_definition = yaml.safe_load(package_definition_io.read())
-    close_files()
+    Returns:
+        package_contents: A dictionary representation of the package yaml
+
+    Raises:
+        InvalidPackage: Raised when the package.yaml is not a regular file
+        ParseError: Raised when there is an error parsing the package yaml
```

Comment:
I'm not clear on the distinction between `ParseError`  (which I assume is actually `ParserError`) and `YAMLError`.  I think the former is just a variation of the latter and you can just handle `YAMLError` and be fine.  Based on the documentation it seems like the intention is that `YAMLError` is the catch-all public facing exception that users are expected to handle.

Reply:
I originally had separate except blocks before I consolidated them into one. The `ParseError` was just a leftover from that implementation